build_utils/bump_changelog.py

```diff
@@ -0,0 +1,115 @@
+#! /usr/bin/env python
+
+"""Bump the current version number in CHANGELOG.rst
+
+Unreleased changes will be moved into a section for the new version number
+"""
+from argparse import ArgumentParser
+from dataclasses import dataclass
+from pathlib import Path
+import re
+
+from toolz.itertoolz import partition
+
+REPOSITORY_ADDRESS = "https://github.com/pace-neutrons/Euphonic"
+
+
+@dataclass
+class Block:
+    """CHANGELOG data corresponding to a single release"""
+    tag: str
+    previous_tag: str
+    content: str
+
+
+def parse_changelog(changelog_file: Path) -> list[Block]:
+    """Read all sections from changelog file
+
+    Each section should begin with two lines in format
+
+      `TAG <repository_address/compare/PREV_TAG...TAG>`_
+      --------------------------------------------------
+
+    except for 'Unreleased' which compares with HEAD
+    """
+
+    with open(changelog_file, "rt", encoding="utf8") as fd:
+        split_text = re.split(r"`(\S+) <\S+/compare/(\S+)\.\.\.\S+>`_\n-+", fd.read())
```

Comment:
Huh, never realised that `re.split` split out the captured groups. That's good to know in future.

Reply:
To avoid the next line, could also do:
```suggestion
        _, *split_text = re.split(r"`(\S+) <\S+/compare/(\S+)\.\.\.\S+>`_\n-+", fd.read())
```
However, this may be pretty arcane and less clear.

Reply:
It's a nice trick but I think a bit misleading here. The first (empty) item _is_ probably something to do with the split text, whereas this implies it is some other return item.